thavalon-webapp/src/components/gameComponents/gameHeader.tsx

```diff
@@ -0,0 +1,115 @@
+ */
+export function GameHeader(): JSX.Element {
+    // State for the current phase of the game
+    const [gamePhaseHeader, setGamePhaseHeader] = useState(GamePhaseHeader.ProposalOther);
+    // State for the number of turns until force. Initialized to 1 to prevent any force-specific actions from occurring.
+    const [turnsUntilForce, setTurnsUntilForce] = useState(1);
+    // State to determine the name of this player.
+    const [me, setMe] = useState("");
+
+    useEffect(() => {
+        const connection = GameSocket.getInstance();
+        connection.onLobbyEvent.subscribe(handleMessage);
+        connection.onGameEvent.subscribe(handleMessage);
+        return () => {
+            connection.onLobbyEvent.unsubscribe(handleMessage);
+            connection.onGameEvent.unsubscribe(handleMessage);
+        }
+    });
+
+    /**
+     * Handles any message received from the server.
+     * @param message An inbound message from the server
+     */
+    function handleMessage(message: InboundMessage): void {
+        switch (message.messageType) {
+            // If it's a snapshot, check the last message to get the current phase.
+            case InboundMessageType.Snapshot:
+                const snapshot = message.data as Snapshot;
+                setMe(snapshot.me);
+                const lastLogIndex = snapshot.log.length - 1;
+                const lastMessage = snapshot.log[lastLogIndex];
+                mapMessageToState(lastMessage);
+                // The last message will always give us the state, but to figure out
+                // the number of proposals remaining, we need a NextProposal message.
+                // Loop backwards over the messages until we find one.
+                if (lastMessage.messageType !== GameMessageType.NextProposal) {
+                    // We only hit this if the last messageType isn't a NextProposal,
+                    // so start the loop from the second to last element.
+                    for (let i = lastLogIndex - 1; i >= 0; i--) {
+                        const logMessage = snapshot.log[i];
+                        if (logMessage.messageType === GameMessageType.NextProposal) {
+                            const proposalMessage = logMessage.data as NextProposalMessage;
+                            setTurnsUntilForce(proposalMessage.max_proposals - proposalMessage.proposals_made);
+                            break;
+                        }
+                    }
+                }
+                break;
+            // If it's a normal game message, just pass in the data to determine the phase.
+            case InboundMessageType.GameMessage:
+                mapMessageToState(message.data as GameMessage);
+                break;
+        }
+    }
+
+    /**
+     * Given a message, determines which phase of the game the player is currently in.
+     * @param message A GameMessage from the server
+     */
+    function mapMessageToState(message: GameMessage): void {
```

Comment:
Up to you, but this could be a good fit for the `useReducer` [hook](https://reactjs.org/docs/hooks-reference.html#usereducer), especially if the state we track gets more complicated. Game messages ought to map pretty directly to actions for the reducer, and we could handle reconnecting by feeding the whole snapshot log to it. When I was setting up the game messages initially, it was pretty heavily inspired by using [Redux](https://redux.js.org/) with React, and `useReducer` looks like a built-in lightweight version.

Reply:
I already have changed this function a lot in a more recent branch, so I'm going to leave it off here. However, `useReducer` looks like it could be quite useful, and I may change to using it for the game phase state changes we have. Already the code I have for that is a pile of spaghet, so this might help solve that issue.